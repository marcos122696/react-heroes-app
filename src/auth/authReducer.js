import { types } from "../types/types";

// const state = {
//     name: 'Marcos',
//     logged: true
// };

// const action = {
//     type: type.login,
//     payload: {
//         name: 'Marcos',
//         email: 'marcos@gmail.com',
//     },
// };

export const authReducer = ( state = {}, action ) => {

    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload, // = name: action.payload.name
                logged: true
            }
        
        case types.logout:
            return {
                logged: false
            };
    
        default:
            return state;
    }
};