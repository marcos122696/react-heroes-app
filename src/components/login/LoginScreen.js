import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const LoginScreen = () => {

    const navigate = useNavigate();
    const { dispatch } = useContext( AuthContext );

    const handleLogin = ()  => {
        const action = {
            type: types.login,
            payload: {
                name: 'Marcos'
            }
        }
        
        dispatch( action );

        const lastPath = localStorage.getItem('lastPath') || '/marvel';

        navigate(lastPath, {
            replace: true
        });
    };

  return (
    <div className="container mt-5 border border-secondary p-5 rounded-pill text-center w-50">
        <h1>Login</h1>
        <hr />

        <button
            className="btn btn-outline-primary col-6"
            onClick={ handleLogin }
        >
            Login
        </button>
    </div>
  )
}
