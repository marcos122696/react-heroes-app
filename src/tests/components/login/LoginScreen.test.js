import { mount } from "enzyme";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { types } from "../../../types/types";


const mockNavigate = jest.fn();
const dispatch = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate
}))

describe('Pruebas en el <LoginScreen />', () => {

    const contextValue = {
        user: {
            logged: false,
        },
        dispatch: dispatch
    }

    const wrapper = mount( 
        <AuthContext.Provider value={ contextValue } >
            <MemoryRouter initialEntries={['/login']}>
                <LoginScreen /> 
            </MemoryRouter>
        </AuthContext.Provider>
    );
    test('Debe mostrarse correctamente', () => { 

        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de realizar el dispatch y la navegacion', () => {

        const action = {
            type: types.login,
            payload: {
                name: 'Marcos'
            }
        };
        
        // localStorage.setItem('lastPath', '/');
        wrapper.find('button').simulate('click');
        expect( dispatch ).toHaveBeenCalledWith( action );
        expect( mockNavigate ).toHaveBeenCalledWith('/marvel', {
            replace: true
        });

        localStorage.setItem('lastPath', '/dc');
        wrapper.find('button').simulate('click');
        expect( dispatch ).toHaveBeenCalledTimes( 2 );
        expect( mockNavigate ).toHaveBeenCalledWith('/dc', {
            replace: true
        });
    });
});