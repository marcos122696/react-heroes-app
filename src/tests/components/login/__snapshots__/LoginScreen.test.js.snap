// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <LoginScreen /> Debe mostrarse correctamente 1`] = `
<div
  className="container mt-5 border border-secondary p-5 rounded-pill"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
